REGRESSION(281488@main): [WPE][GTK] Process launching is slow

mcatanzaro opened the pull request:
#### cf8f5f540b30716eb35ed55b7ea1fc6ca78534c1
<pre>
REGRESSION(281488@main): [WPE][GTK] Process launching is slow
<a href="https://bugs.webkit.org/show_bug.cgi?id=279026">https://bugs.webkit.org/show_bug.cgi?id=279026</a>

Reviewed by Carlos Garcia Campos and Adrian Perez de Castro.

ProcessLauncher is already designed to support asynchronous process
launching, but the GLib implementation ignores this and does everything
synchronously. Previously this was OK because
ProcessLauncher::launchProcess would return immediately after the
subprocess is spawned, without waiting for any code to execute in the
subprocess. Apparently that&apos;s fast enough in practice. But after my
changes in 281488@main, we now additionally wait for the subprocess to
send credentials to the parent via a socket. That turns out to be
drastically slower than actually spawning the process.

The blocking is unnecessary because the code was already carefully
structured to allow asynchronicity. I just failed to take advantage of
it.

Instead, return immediately after spawning the subprocess. Then, wait
until the socket is ready before reading the pid from it. Now the read
can complete instantaneously instead of blocking the UI process.

* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: <a href="https://commits.webkit.org/283414@main">https://commits.webkit.org/283414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57684499be6f0e67461b7958db031759f4d56798

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66239 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18858 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70271 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16849 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17130 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53145 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11744 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69306 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/42070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57351 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33787 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/38740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15725 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/60637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15076 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71974 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10194 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14466 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60466 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10226 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57415 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60768 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14595 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8419 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41420 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42496 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43679 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42240 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -235,21 +235,37 @@ void ProcessLauncher::launchProcess()
         g_error("Unable to spawn a new child process: %s", error->message);
 
 #if OS(LINUX)
-    m_processID = IPC::readPIDFromPeer(pidSocketPair.server);
-    RELEASE_ASSERT(!close(pidSocketPair.server));
```
Now that I look at this, why do we do this always on linux? Shouldn't we do this only when using bwrap sandbox?

aperezdc replied:
I suppose it's to keep only one code path in all cases, because this will always work regardless of using sandboxing or not, or which kind of sandbox. But's let's wait for @mcatanzaro to confirm.

mcatanzaro replied:
Adrian is correct. This is needed for bubblewrap sandbox and also flatpak sandbox, but not when running unsandboxed. It's harmless to do always, though. If we want to omit it when running unsandboxed, then the code would become a little more complicated; specifically, the auxiliary process would have a different number of command line arguments depending on whether we are sandboxed or not.

aperezdc commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -235,21 +235,37 @@ void ProcessLauncher::launchProcess()
+    g_socket_set_blocking(pidSocket.get(), FALSE);
+    m_socketMonitor.start(pidSocket.get(), G_IO_IN, RunLoop::main(), [protectedThis = Ref { *this }, this, pidSocket, serverSocket = webkitSocketPair.server](GIOCondition condition) -> gboolean {
+        if (condition & G_IO_ERR || condition & G_IO_NVAL)
+            g_error("Failed to read pid from child process");
```
We want to handle `G_IO_HUP` here as well to detect whether the socket pair has been closed prematurely on the other end. We may want to add `safeStrerror(errno)` to the error message as well, in case it may help troubleshoot issues, may they arise.

mcatanzaro replied:
I actually tried that, but it didn't work. The problem is the child process is going to send its message and then immediately close the socket, so `G_IO_HUP` is one of the events that gets received here normally.

That said, maybe I should detect `(condition & G_IO_HUP) && !(condition & G_IO_IN)` as an error. That shouldn't ever happen, but if it does, the ProcessLauncher will just get stuck forever without failing.

> We may want to add safeStrerror(errno) to the error message as well, in case it may help troubleshoot issues, may they arise.

Hm, I'm not sure, because we're not directly calling any function that sets errno.

mcatanzaro replied:
> That said, maybe I should detect (condition & G_IO_HUP) && !(condition & G_IO_IN) as an error. That shouldn't ever happen, but if it does, the ProcessLauncher will just get stuck forever without failing.

(I'm not sure whether this can actually happen or not.)

mcatanzaro replied:
> That said, maybe I should detect (condition & G_IO_HUP) && !(condition & G_IO_IN) as an error. That shouldn't ever happen, but if it does, the ProcessLauncher will just get stuck forever without failing.

Looking over the `poll(2)` man page, I think this is correct. For our purposes, it's an error only if it occurs on its own.

mcatanzaro replied:
Updated. This look OK? Please assign to merge queue if so.

carlosgcampos replied:
Maybe it would be simpler to first check `condition & G_IO_IN` and then you only need to check the others after the return

mcatanzaro replied:
Good point. Actually, I think this would work:

```
if (!(condition & G_IO_IN))
    g_error("Failed to read pid from child process");
```

aperezdc reviewed (approved):
Nice that we can make process launching fully async. Probably it would have been good to have even before the recent change to pass real PIDs to the launcher&mdash;better late than never!

mcatanzaro committed:
REGRESSION(281488@main): [WPE][GTK] Process launching is slow
https://bugs.webkit.org/show_bug.cgi?id=279026

Reviewed by Carlos Garcia Campos and Adrian Perez de Castro.

ProcessLauncher is already designed to support asynchronous process
launching, but the GLib implementation ignores this and does everything
synchronously. Previously this was OK because
ProcessLauncher::launchProcess would return immediately after the
subprocess is spawned, without waiting for any code to execute in the
subprocess. Apparently that's fast enough in practice. But after my
changes in 281488@main, we now additionally wait for the subprocess to
send credentials to the parent via a socket. That turns out to be
drastically slower than actually spawning the process.

The blocking is unnecessary because the code was already carefully
structured to allow asynchronicity. I just failed to take advantage of
it.

Instead, return immediately after spawning the subprocess. Then, wait
until the socket is ready before reading the pid from it. Now the read
can complete instantaneously instead of blocking the UI process.

* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: https://commits.webkit.org/283414@main
